packages/devtools_app/lib/src/shared/common_widgets.dart

```diff
@@ -1421,6 +1427,74 @@ class FormattedJson extends StatelessWidget {
   }
 }
 
+class JsonViewer extends StatefulWidget {
+  const JsonViewer({
+    required this.json,
+  });
+
+  final String json;
+
+  @override
+  State<JsonViewer> createState() => _JsonViewerState();
+}
+
+class _JsonViewerState extends State<JsonViewer> {
+  late Future<void> _initializeTree;
+  late DartObjectNode variable;
+
+  @override
+  void initState() {
+    super.initState();
+    assert(widget.json.isNotEmpty);
+    final responseJson = json.decode(widget.json);
+    // Insert the JSON data into the fake service cache so we can use it with
+    // the `ExpandableVariable` widget.
+    final root =
+        serviceManager.service!.fakeServiceCache.insertJsonObject(responseJson);
+    variable = DartObjectNode.fromValue(
+      name: '[root]',
+      value: root,
+      artificialName: true,
+      isolateRef: IsolateRef(
+        id: 'fake-isolate',
+        number: 'fake-isolate',
+        name: 'local-cache',
+        isSystemIsolate: true,
+      ),
+    );
+    _initializeTree = buildVariablesTree(variable);
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    // Remove the JSON object from the fake service cache to avoid holding on
+    // to large objects indefinitely.
+    serviceManager.service!.fakeServiceCache
+        .removeJsonObject(variable.value as Instance);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+      padding: const EdgeInsets.all(denseSpacing),
+      child: SingleChildScrollView(
+        child: FutureBuilder(
+          future: _initializeTree,
+          builder: (context, snapshot) {
+            if (snapshot.connectionState != ConnectionState.done)
+              return Container();
+            return ExpandableVariable(
+              variable: variable,
+              debuggerController: Provider.of<DebuggerController>(context),
```

Comment:
pull the debugger controller in didChangeDependencies so we don't have to look it up from package:provider on each build. Better yet, use the new `ProvidedControllerMixin` on this state class.